### PR TITLE
Make `conda env update --prune` ignore history specs.

### LINKED ISSUE #7279: conda env update --prune does not remove installed packages not defined in environment.yml
**I'm submitting a...**
  - [x] bug report
  - [ ] feature request

## Current Behavior
`conda env update --prune` does not remove packages no longer referenced in the `environment.yml` file.

### Steps to Reproduce

Define an `environment.yml` file in a project directory as follows:

    name: new-project-env
    dependencies:
    - python=3
    - numpy

In the project directory run the following commands:

    conda env create
    conda activate new-project-env
    python
    import numpy
    exit()

This succeeds as expected.
Update the `environment.yml` file as follows, **removing numpy**,

    name: new-project-env
    dependencies:
    - python=3

Run the following commands in the project directory:

    conda deactivate
    conda env update --prune
    conda activate new-project-env
    python
    import numpy
    exit()

**This numpy import succeeds and seems like a bug**. 

## Expected Behavior
Per [the docs](https://conda.io/docs/commands/env/conda-env-update.html) (--prune means "remove installed packages not defined in environment.yml") I expect that after installing a package (numpy in the example above) and subsquently removing it from `environment.yml`  the package should be removed from the relevant conda environment.

## Current Workaround
I've found that we can use `conda env create --force` to overwrite the existing environment and thus ensure that the environment accurately reflects the `environment.yml` file. But this requires recreating the environment from scratch.

## Environment Information
<details open><summary><code>`conda info`</code></summary><p>

```
     active environment : new-project-env
    active env location : /home/mikhailgolubitsky/anaconda3/envs/new-project-env
            shell level : 2
       user config file : /home/mikhailgolubitsky/.condarc
 populated config files : /home/mikhailgolubitsky/.condarc
          conda version : 4.5.3
    conda-build version : 3.0.27
         python version : 3.6.3.final.0
       base environment : /home/mikhailgolubitsky/anaconda3  (writable)
           channel URLs : https://conda.anaconda.org/conda-forge/linux-64
                          https://conda.anaconda.org/conda-forge/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/free/linux-64
                          https://repo.anaconda.com/pkgs/free/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
                          https://repo.anaconda.com/pkgs/pro/linux-64
                          https://repo.anaconda.com/pkgs/pro/noarch
          package cache : /home/mikhailgolubitsky/anaconda3/pkgs
                          /home/mikhailgolubitsky/.conda/pkgs
       envs directories : /home/mikhailgolubitsky/anaconda3/envs
                          /home/mikhailgolubitsky/.conda/envs
               platform : linux-64
             user-agent : conda/4.5.3 requests/2.18.4 CPython/3.6.3 Linux/4.13.0-39-generic ubuntu/17.10 glibc/2.26
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```
</p></details>

<details open><summary><code>`conda config --show-sources`</code></summary><p>

```
==> /home/mikhailgolubitsky/.condarc <==
channels:
  - conda-forge
  - defaults
```
</p></details>

<details><summary><code>`conda list --show-channel-urls`</code></summary><p>

```
# packages in environment at /home/mikhailgolubitsky/anaconda3/envs/new-project-env:
#
# Name                    Version                   Build  Channel
ca-certificates           2018.4.16                     0    conda-forge
certifi                   2018.4.16                py36_0    conda-forge
intel-openmp              2018.0.0                      8    defaults
libgcc-ng                 7.2.0                hdf63c60_3    defaults
libgfortran-ng            7.2.0                hdf63c60_3    defaults
mkl                       2018.0.2                      1    defaults
mkl_fft                   1.0.2                    py36_0    conda-forge
mkl_random                1.0.1                    py36_0    conda-forge
ncurses                   5.9                          10    conda-forge
numpy                     1.14.3           py36h14a74c5_0    defaults
numpy-base                1.14.3           py36hdbf6ddf_0    defaults
openssl                   1.0.2o                        0    conda-forge
pip                       9.0.3                    py36_0    conda-forge
python                    3.6.5                         1    conda-forge
readline                  7.0                           0    conda-forge
setuptools                39.1.0                   py36_0    conda-forge
sqlite                    3.20.1                        2    conda-forge
tk                        8.6.7                         0    conda-forge
wheel                     0.31.0                   py36_0    conda-forge
xz                        5.2.3                         0    conda-forge
zlib                      1.2.11                        0    conda-forge


```
</p></details>

### PR DESCRIPTION
**Pull request by tarcisioe**
This is supposed to solve #7279. The idea here is that if we are pruning we must use only the specs coming from environment.yml and ignore the history. Regression tests have also been added.

Resolves #7279

### CODE REVIEW

**Comment by nicoddemus** [comment]
Great work @tarcisioe! Fixing `--prune` means we can upgrade our conda version at @ESSS. 👍

File: tests/core/test_solve.py [hunk 1]
```diff
@@ -363,6 +363,153 @@ def test_prune_1(tmpdir):
...
+            'channel-1::numexpr-2.1-np17py27_p0',
+            'channel-1::scipy-0.12.0-np17py27_p0',
+            'channel-1::numbapro-0.11.0-np17py27_p0',
+            'channel-1::scikit-learn-0.13.1-np17py27_p0',
+            'channel-1::mkl-11.0-np17py27_p0',
+            'channel-1::accelerate-1.1.0-np17py27_p0',
+        ))
+        assert convert_to_dist_str(final_state_1) == order
+
+    new_environment_specs = MatchSpec("numpy"), MatchSpec("python=2.7.3"),
```
Comment (nicoddemus) [inline tests/core/test_solve.py]: I think here would be a good place to comment what's the relation between the new specs and the test intent according to the docstring (same applies to the other tests).

**Comment by tarcisioe** [comment]
@marcelotrevisani @msarahan I have submitted my CLA (twice, actually), is anything else needed for the bot to see it? Maybe it needs to be updated in .cla-signers?

**Comment by tarcisioe** [comment]
Also, I believe appveyor/pr wasn't broken by my change, and I'm not too sure about what's happened in circleci conda-build. Should I do something about those?

**Comment by marcelotrevisani** [comment]
> @marcelotrevisani @msarahan I have submitted my CLA (twice, actually), is anything else needed for the bot to see it? Maybe it needs to be updated in .cla-signers?

Hi @tarcisioe , this process has a manual part lets say. @msarahan has to review and approve it. It seems to be a legal process which they cannot avoid.

File: tests/core/test_solve.py [hunk 2]
```diff
@@ -363,6 +363,170 @@ def test_prune_1(tmpdir):
         assert convert_to_dist_str(link_precs) == link_order
 
 
+def test_update_prune_1(tmpdir):
```
Comment (marcelotrevisani) [inline tests/core/test_solve.py]: Thanks for the tests! :)

**Comment by marcelotrevisani** [comment]
@cla-bot check

**Comment by kahing** [comment]
what's the status of this? does the maintainer agree that this is the right fix?

**Comment by kahing** [comment]
I tried to manually apply this patch to 4.8.2 and the steps in https://github.com/conda/conda/issues/7279#issuecomment-583170554 is still broken

**Comment by tarcisioe** [comment]
@kahing It hasn't been reviewed yet by the maintainers. I will try to reproduce the steps you posted and investigate.

**Comment by tarcisioe** [comment]
Well, shame on me I guess. I developed this patch on top of 4.7.x and trusted my regression tests. They are indeed still passing, so something __else__ reintroduced the issue. I went through the major tags rebasing my patch onto them:

- 4.7.9 was ok
- 4.8.0 was ok
- 4.8.1 stopped working

So I made a script that applied my patch and tested, and got the change that ended up nullifying my fix with `git bisect`. It was commit 03d22d4. I will work on matching my fix to what was intended on that commit now.

**Comment by tarcisioe** [comment]
@kahing it's now rebased on master and when pruning the option "freeze-installed" is ignored (which was making packages that should otherwise be pruned be kept). I'm unsure if that is the best solution, but it basically reverts prune to the previous behaviour. Another regression test was added.

**Comment by kalefranz** [comment]
Thanks for contributing this PR, and apologies for the delay in review. We're undergoing some changes in the Conda team right now, but it's our plan to be spending considerably more resources directly on Conda soon.

In particular for this PR, does this PR change the behavior of `conda install --prune` or `conda update --prune`?

**Comment by tarcisioe** [comment]
@kalefranz It shouldn't. What it does is:

- When prune is enabled, history is not included in the specs (since that ends up including every package in the spec and therefore avoids any package being pruned). 
- Disables `FREEZE_INSTALLED` when pruning. This was done mostly because I wasn't able to make both things work properly together, but I considered it reasonable since the change that introduced that modifier as a default (03d22d4) was made in between 4.8.0 and 4.8.1, when prune was already not working.

Therefore, `--prune --freeze-installed` will ignore `--freeze-installed`, but I am unsure if those flags ever worked together.

Also, no issues with the delay, I am aware of the size of conda as a project and how much work you guys have. I'm the one to thank you for the maintenance!

**Comment by prusse-martin** [comment]
WIP
The current state can trigger an error due the missing installed specs.

**Comment by prusse-martin** [comment]
The detected problem in the PR have been fixed.
I did toil a bit with the master branch in my fork but was unable to get a job to succeed.
Some integration tests are still failing there.

**Comment by nicoddemus** [comment]
We have rebase the PR over the latest `master`, and this should be ready for another review/merge.

cc @marcelotrevisani

**Comment by marcelotrevisani** [comment]
pinging @chenghlee  as well

**Comment by nicoddemus** [comment]
I see @tarcisioe has not signed the CLA, is this what is holding back this PR from being reviewed/merged?

@tarcisioe could you please try again to sign the agreement?

**Comment by rcasero** [comment]
@tarcisioe this PR seemed to be very close to success, but was held back by a failed test in windows. Any chance you could complete it? It'd be very useful

**Comment by rcasero** [comment]
@jezdez pushed this 3 months ago, and @nicoddemus a minute ago, but it still seems to be a problem in conda 4.14.0. Any version where this is fixed?

**Comment by nicoddemus** [comment]
> @jezdez pushed this 3 months ago, and @nicoddemus a minute ago, but it still seems to be a problem in conda 4.14.0. 

As the PR is still open, it means it did not go into any version yet... hopefully if it is merged it will then be shipped in the next conda release.

**Comment by jezdez** [comment]
(restarting CI which was stuck on Codecov service issues)

**Comment by rcasero** [comment]
@jezdez the macos CI tests may be stuck waiting for a runner.

**Comment by jezdez** [comment]
> In those cases it is better to re-run just the failed jobs (or maintainers can choose to ignore those too... unfortunately codecov is not very reliable and we often see those failures in pytest).

You're right and I tried that first, but for some reason GitHub's UI didn't offer that option for those jobs. Not sure why.

**Comment by jezdez** [comment]
> @jezdez the macos CI tests may be stuck waiting for a runner.

Possibly, we've yet to increase the number of runners for macOS and in some scenarios we simply use up the available runners, especially given more activity in projects like constructor.

**Comment by rcasero** [comment]
@jezdez All checks have passed. Is that good enough for you guys to review and merge?

**Comment by andreostrovsky** [comment]
@kalefranz @jezdez could this pr please just get reviewed, it's been waiting for over 3 months. This env update functionality has been broken for over 4.5 years now.

**Comment by karansampath** [comment]
Bumping on whether this PR has been merged?

File: conda/core/solve.py [hunk 3]
```diff
@@ -1341,8 +1341,9 @@ def __init__(
         self.r = None
 
         # Group 4. Mutable working containers
-        self.specs_map = {}
-        self.solution_precs = tuple(self.prefix_data.iter_records())
+        self.specs_map = odict()
```
Comment (beeankha) [inline conda/core/solve.py:1344]: We've removed usage of `odict` in conda (see https://github.com/conda/conda/pull/12254)
```suggestion
        self.specs_map = {}
```

File: conda/core/solve.py [hunk 4]
```diff
@@ -1365,9 +1366,16 @@ def pinned_specs(self):
...
+        if self.prune:
+            # DO NOT add existing prefix data to solution on prune
+            self.solution_precs = tuple()
+        else:
+            self.solution_precs = tuple(self.prefix_data.iter_records())
+
     def working_state_reset(self):
-        self.specs_map = {}
-        self.solution_precs = tuple(self.prefix_data.iter_records())
+        self.specs_map = odict()
```
Comment (beeankha) [inline conda/core/solve.py:1377]:
```suggestion
        self.specs_map = {}
```

File: conda/core/solve.py [hunk 5]
```diff
@@ -520,54 +522,53 @@ def empty_package_list(pkg):
...
-                or MatchSpec(prec.name) in context.aggressive_update_packages
-                or prec.subdir == "pypi"
-            ):
-                ssc.specs_map.update({prec.name: MatchSpec(prec.name)})
+        if ssc.prune:
+            # When pruning DO NOT consider history of already installed packages when solving.
+            prepared_specs = set(concatv(
+                self.specs_to_remove,
+                self.specs_to_add,
+            ))
```
Comment (beeankha) [inline conda/core/solve.py:530]: We've also removed usage of `concatv()` (please see https://github.com/conda/conda/pull/12020)

```suggestion
            prepared_specs = set(*self.specs_to_remove, *self.specs_to_add)
```

File: conda/core/solve.py [hunk 6]
```diff
@@ -520,54 +522,53 @@ def empty_package_list(pkg):
...
+                if (not ssc.specs_from_history_map
+                        or MatchSpec(prec.name) in context.aggressive_update_packages
+                        or prec.subdir == 'pypi'):
+                    ssc.specs_map.update({prec.name: MatchSpec(prec.name)})
+
+            prepared_specs = set(concatv(
+                self.specs_to_remove,
+                self.specs_to_add,
+                ssc.specs_from_history_map.values(),
+            ))
```
Comment (beeankha) [inline conda/core/solve.py:571]:
```suggestion
            prepared_specs = set(
                *self.specs_to_remove,
                *self.specs_to_add,
                *ssc.specs_from_history_map.values(),
            )
```

File: conda/core/solve.py [hunk 7]
```diff
@@ -731,17 +732,16 @@ def _add_specs(self, ssc):
         # the only things we should consider freezing are things that don't conflict with the new
         #    specs being added.
         explicit_pool = ssc.r._get_package_pool(self.specs_to_add)
+        if ssc.prune:
+            # Ignore installed specs on prune.
+            installed_specs = ()
+        else:
+            installed_specs = tuple(concatv(
+                (_.to_match_spec() for _ in ssc.prefix_data.iter_records())
+            ))
```
Comment (beeankha) [inline conda/core/solve.py:741]:
```suggestion
            installed_specs = tuple(
                *(_.to_match_spec() for _ in ssc.prefix_data.iter_records())
            )
```

File: conda/core/solve.py (hunk 3 above)
Comment (dholth) [reply conda/core/solve.py:1344]: (since ordinary Python dicts are ordered now)

File: conda/core/solve.py (hunk 7 above)
Comment (kenodegard) [reply conda/core/solve.py:741]: not quite, i think you need to use [`itertools.chain.from_iterable`](https://docs.python.org/3/library/itertools.html#itertools.chain.from_iterable) instead

**Comment by beeankha** [comment]
pre-commit.ci autofix

File: conda/core/solve.py [hunk 8]
```diff
@@ -734,11 +744,15 @@ def _add_specs(self, ssc):
             installed_specs = ()
         else:
             installed_specs = tuple(
-                chain.from_iterable(_.to_match_spec() for _ in ssc.prefix_data.iter_records())
+                chain.from_iterable(
+                    _.to_match_spec() for _ in ssc.prefix_data.iter_records()
```
Comment (kenodegard) [inline conda/core/solve.py:748]:
```suggestion
                    record.to_match_spec() for record in ssc.prefix_data.iter_records()
```

File: conda/core/solve.py [hunk 9]
```diff
@@ -734,11 +744,15 @@ def _add_specs(self, ssc):
...
+                )
             )
 
-        conflict_specs = ssc.r.get_conflicting_specs(installed_specs, self.specs_to_add) or tuple()
-        conflict_specs = set(_.name for _ in conflict_specs)
+        conflict_specs = (
+            ssc.r.get_conflicting_specs(installed_specs, self.specs_to_add) or tuple()
+        )
+        conflict_specs = {_.name for _ in conflict_specs}
```
Comment (kenodegard) [inline conda/core/solve.py:755]:
```suggestion
        conflict_specs = {spec.name for spec in conflict_specs}
```

File: tests/core/test_solve.py [hunk 10]
```diff
@@ -434,162 +434,190 @@ def test_update_prune_1(tmpdir):
...
-    new_environment_specs = MatchSpec("numpy"), MatchSpec("python=2.7.3"),
+    new_environment_specs = (
+        MatchSpec("numpy"),
+        MatchSpec("python=2.7.3"),
+    )
 
-    with get_solver(tmpdir, new_environment_specs, prefix_records=final_state_1,
-                    history_specs=specs) as solver:
+    with get_solver(
+        tmpdir, new_environment_specs, prefix_records=final_state_1, history_specs=specs
```
Comment (kenodegard) [inline tests/core/test_solve.py:602]:
```suggestion
        tmpdir,
        new_environment_specs,
        prefix_records=final_state_1,
        history_specs=specs,
```

File: tests/core/test_solve.py [hunk 11]
```diff
@@ -598,66 +626,78 @@ def test_update_prune_4(tmpdir):
...
     # Here numpy is removed from the specs (but the old specs are kept as
     # history).
-    new_environment_specs = MatchSpec("python=2.7.3"),
+    new_environment_specs = (MatchSpec("python=2.7.3"),)
 
-    with get_solver(tmpdir, new_environment_specs, prefix_records=final_state_1,
-                    history_specs=specs) as solver:
-        final_state_1 = solver.solve_final_state(prune=True, update_modifier=UpdateModifier.FREEZE_INSTALLED)
+    with get_solver(
+        tmpdir, new_environment_specs, prefix_records=final_state_1, history_specs=specs
```
Comment (kenodegard) [inline tests/core/test_solve.py:656]:
```suggestion
        tmpdir,
        new_environment_specs,
        prefix_records=final_state_1,
        history_specs=specs,
```

File: tests/core/test_solve.py [hunk 12]
```diff
@@ -598,66 +626,78 @@ def test_update_prune_4(tmpdir):
...
+    ) as solver:
+        final_state_1 = solver.solve_final_state(
+            prune=True, update_modifier=UpdateModifier.FREEZE_INSTALLED
```
Comment (kenodegard) [inline tests/core/test_solve.py:659]:
```suggestion
            prune=True,
            update_modifier=UpdateModifier.FREEZE_INSTALLED,
```

File: tests/core/test_solve.py [hunk 13]
```diff
@@ -598,66 +626,78 @@ def test_update_prune_4(tmpdir):
...
-    assert 'Updating numexpr is constricted by' not in out
+    assert "Updating numexpr is constricted by" not in out
 
     # If prefix data is evaluated it will conflict with "pinned" dependencies of new specs.
-    new_environment_specs = MatchSpec("python=2.7"), MatchSpec("numexpr==2.0.1=np17py27_p2")
-    with get_solver(tmpdir, new_environment_specs, prefix_records=final_state_1,
-                    history_specs=specs) as solver:
+    new_environment_specs = MatchSpec("python=2.7"), MatchSpec(
+        "numexpr==2.0.1=np17py27_p2"
+    )
```
Comment (kenodegard) [inline tests/core/test_solve.py:692]:
```suggestion
    new_environment_specs = (
        MatchSpec("python=2.7"),
        MatchSpec("numexpr==2.0.1=np17py27_p2"),
    )
```

File: tests/core/test_solve.py [hunk 14]
```diff
@@ -598,66 +626,78 @@ def test_update_prune_4(tmpdir):
...
+    with get_solver(
+        tmpdir, new_environment_specs, prefix_records=final_state_1, history_specs=specs
```
Comment (kenodegard) [inline tests/core/test_solve.py:694]:
```suggestion
        tmpdir,
        new_environment_specs,
        prefix_records=final_state_1,
        history_specs=specs,
```

File: tests/core/test_solve.py [hunk 15]
```diff
@@ -429,6 +429,277 @@ def test_prune_1(tmpdir):
...
+        )
+        assert convert_to_dist_str(final_state_1) == order
+
+    # Here accelerate is removed from the specs (but the old specs are kept as
+    # history). This should cause every dependency of accelerate to disappear
+    # from the pruned package list.
+    new_environment_specs = (MatchSpec("python=2.7.3"),)
+
+    with get_solver(
+        tmpdir, new_environment_specs, prefix_records=final_state_1, history_specs=specs
```
Comment (kenodegard) [inline tests/core/test_solve.py:530]:
```suggestion
        tmpdir,
        new_environment_specs,
        prefix_records=final_state_1,
        history_specs=specs,
```

File: tests/core/test_solve.py [hunk 16]
```diff
@@ -429,6 +429,277 @@ def test_prune_1(tmpdir):
...
+            )
+        )
+        assert convert_to_dist_str(final_state_1) == order
+
+    # Here numpy is removed from the specs (but the old specs are kept as
+    # history).
+    new_environment_specs = (MatchSpec("python=2.7.3"),)
+
+    with get_solver(
+        tmpdir, new_environment_specs, prefix_records=final_state_1, history_specs=specs
```
Comment (kenodegard) [inline tests/core/test_solve.py:464]:
```suggestion
        tmpdir,
        new_environment_specs,
        prefix_records=final_state_1,
        history_specs=specs,
```

File: tests/core/test_solve.py [hunk 17]
```diff
@@ -429,6 +429,277 @@ def test_prune_1(tmpdir):
...
+            )
+        )
+        assert convert_to_dist_str(final_state_1) == order
+
+
+@pytest.mark.parametrize("prune", [True, False])
+def test_update_prune_5(tmpdir, prune, capsys):
+    """Regression test: Check that prefix data is not taken into account when solving on prune."""
+    # "Create" a conda env with specs that "pin" dependencies.
+    specs = MatchSpec("python=2.7"), MatchSpec("numexpr==2.0.1=np17py27_p3")
```
Comment (kenodegard) [inline tests/core/test_solve.py:682]:
```suggestion
    specs = (
        MatchSpec("python=2.7"),
        MatchSpec("numexpr==2.0.1=np17py27_p3"),
    )
```

File: conda/core/solve.py [hunk 18]
```diff
@@ -731,15 +739,18 @@ def _add_specs(self, ssc):
...
         explicit_pool = ssc.r._get_package_pool(self.specs_to_add)
+        if ssc.prune:
+            # Ignore installed specs on prune.
+            installed_specs = ()
+        else:
+            installed_specs = tuple(
+                chain.from_iterable(
+                    _.to_match_spec() for _ in ssc.prefix_data.iter_records()
+                )
+            )
```
Comment (kenodegard) [inline conda/core/solve.py:750]: Actually this is incorrect, `Record.to_match_spec()` doesn't return an iterable, it only returns a single spec, no chaining is necessary:

```suggestion
            installed_specs = [
                record.to_match_spec() for record in ssc.prefix_data.iter_records()
            ]
```

File: conda/core/solve.py (hunk 8 above)
Comment (kenodegard) [reply conda/core/solve.py:748]: see https://github.com/conda/conda/pull/9614/files#r1298796851

File: conda/core/solve.py (hunk 18 above)
Comment (beeankha) [reply conda/core/solve.py:750]: Thanks for the re-naming, that helps clarifies what's going on 👍

**Review by kenodegard — CHANGES_REQUESTED** [review]
Looks like it is passing 🎉, just need the news snippet and then this will be good to go

**Review by kenodegard — APPROVED** [review]

**Review by beeankha — APPROVED** [review]

**Review by jaimergp — DISMISSED** [review]
TIL that `--prune` is only enabled in `conda env`. I had the impression that it was a general flag (and the code in `solve.py` does not check for the subcommand used), but it's not exposed in any other CLI. :shrug:

For that reason, I'd change the PR title to "Make `conda env update --prune` ignore history specs"? Although the tests seem to be general and could be extended to `conda update` if the CLI was updated.

File: news/9614-prune-ignores-history-specs [hunk 19]
```diff
@@ -0,0 +1,19 @@
+### Enhancements
+
+* <news item>
+
+### Bug fixes
+
+* `update --prune` uses only the specs coming from `environment.yml` file and ignores the history specs. (#9614)
```
Comment (beeankha) [inline news/9614-prune-ignores-history-specs:7]:
```suggestion
* `conda env update --prune` uses only the specs coming from `environment.yml` file and ignores the history specs. (#9614)
```

**Review by beeankha — APPROVED** [review]

**Pull request merged**